assets/js/modules/adsense/components/setup/v2/SetupMain.js

```diff
@@ -0,0 +1,319 @@
+	const [ isSubmittingInBackground, setIsSubmittingInBackground ] = useState(
+		false
+	);
+
+	const siteURL = useSelect( ( select ) =>
+		select( CORE_SITE ).getReferenceSiteURL()
+	);
+	const isAdBlockerActive = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).isAdBlockerActive()
+	);
+	const accounts = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getAccounts()
+	);
+	const previousAccountID = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getAccountID()
+	);
+	const previousClientID = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getClientID()
+	);
+	const accountsError = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getError( 'getAccounts', [] )
+	);
+	// Check whether settings differ from server and are valid.
+	const canSubmitChanges = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).canSubmitChanges()
+	);
+
+	const accountID = determineAccountID( {
+		accounts,
+		previousAccountID,
+	} );
+
+	const clients = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getClients( accountID )
+	);
+	const alerts = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getAlerts( accountID )
+	);
+	const alertsError = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getError( 'getAlerts', [ accountID ] )
+	);
+
+	const clientID = determineClientID( {
+		clients,
+		previousClientID,
+	} );
+	const urlChannels = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getURLChannels( accountID, clientID )
+	);
+	const urlChannelsError = useSelect( ( select ) =>
+		select( MODULES_ADSENSE ).getErrorForSelector( 'getURLChannels', [
+			accountID,
+			clientID,
+		] )
+	);
+
+	const accountStatus = determineAccountStatus( {
+		accounts,
+		clients,
+		alerts,
+		urlChannels,
+		accountsError,
+		alertsError,
+		urlChannelsError,
+		previousAccountID,
+		previousClientID,
+	} );
+	const siteStatus = determineSiteStatus( {
+		urlChannels,
+		siteURL,
+	} );
+
+	useUpdateEffect( () => {
+		if (
+			accountID !== undefined ||
+			clientID !== undefined ||
+			accountStatus !== undefined ||
+			siteStatus !== undefined
+		) {
+			setIsAwaitingBackgroundSubmit( true );
+		}
+	}, [ accountID, clientID, accountStatus, siteStatus ] );
+
+	// Update current account ID setting on-the-fly.
+	useEffect( () => {
+		if (
+			! accountID ||
+			accounts?.length !== 1 ||
+			previousAccountID ||
+			( accounts?.length === 1 &&
+				accounts.findIndex( ( { _id } ) => _id === accountID ) === -1 )
+		) {
+			return;
+		}
+
+		setAccountID( accountID );
+		// Set flag to await background submission.
+		setIsAwaitingBackgroundSubmit( true );
+	}, [ accounts, accountID, previousAccountID, setAccountID ] );
+
+	// Update account status on-the-fly.
+	useEffect( () => {
+		if ( accounts?.length === 0 ) {
+			setAccountStatus( ACCOUNT_STATUS_NONE );
+		} else if ( accounts?.length && ! accountID ) {
+			setAccountStatus( ACCOUNT_STATUS_MULTIPLE );
+		} else {
```

Comment:
We need to make sure that accounts are loaded:
```suggestion
		} else if ( accounts?.length ) {
```

Maybe it is even worth extracting `accounts` check into its own condition, to return early if they are not loaded yet 🤔

Reply:
Actually if `accounts` is not loaded, we are showing the progress bar 👍

Reply:
Thanks, @asvinb.

> Actually if accounts is not loaded, we are showing the progress bar 👍

Yes, but useEffect hook is still executed even if we return a progress bar. So we need to ensure that we don't change the account status before accounts are loaded.